### PR TITLE
Draft: Validation of user email addresses

### PR DESCRIPTION
**Pull request by luelista**

### CODE REVIEW

**Comment by luelista** [comment]
image


image


image


image

File: src/pretix/base/models/auth.py [hunk 1]
```diff
@@ -353,6 +356,43 @@ def send_security_notice(self, messages, email=None):
...
+                'user': self,
+                'reason': msg,
+                'code': code,
+            },
+            event=None,
+            user=self,
+            locale=self.locale
+        )
+
+    def check_confirmation_code(self, reason, code):
```
Comment (luelista) [inline src/pretix/base/models/auth.py]: todo: this method needs a limit on how many times it can be called on the same generated code, to prevent brute force attacks

File: src/pretix/base/models/auth.py [hunk 2]
```diff
@@ -353,6 +356,43 @@ def send_security_notice(self, messages, email=None):
...
+
+        with language(self.locale):
+            if reason == 'email_change':
+                msg = str(_('to confirm changing your email address from {old_email}\nto {new_email}, use the following code:').format(
+                    old_email=self.email, new_email=email,
+                ))
+            else:
+                raise Exception('Invalid confirmation code reason')
+
+        code = "%07d" % random.randint(0, 9999999)
```
Comment (raphaelm) [inline src/pretix/base/models/auth.py]: I'd favor 
```suggestion
        code = "%07d" % secrets.randint(0, 9999999)
```
to make sure a secure RNG is used

File: src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+{% block content %}
+	<form class="form-signin" action="" method="post">
+        <h3>{% trans "Enter confirmation code" %}</h3>
+        {% csrf_token %}
+        <p>{{ message }}</p>
+        {% bootstrap_form_errors form type='all' layout='inline' %}
+        {% bootstrap_field form.code %}
+		<div class="form-group buttons">
+            <button type="submit" class="btn btn-primary">
+                {% trans "Okay" %}
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html]: For consistent UI and ease of translation, I'd favor a button label we already use elsewhere (Continue, Confirm, …)

File: src/pretix/control/templates/pretixcontrol/user/change_email.html [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{% extends "pretixcontrol/base.html" %}
+{% load i18n %}
+{% load bootstrap3 %}
+{% block title %}{% trans "Change login email address" %}{% endblock %}
+{% block content %}
+    <form action="" method="post" class="form centered-form">
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/user/change_email.html:6]: Why does this layout different than any other backend screen?

File: src/pretix/control/templates/pretixcontrol/user/change_password.html [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{% extends "pretixcontrol/base.html" %}
+{% load i18n %}
+{% load bootstrap3 %}
+{% block title %}{% trans "Change password" %}{% endblock %}
+{% block content %}
+    <form action="" method="post" class="form centered-form">
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/user/change_password.html:6]: Why does this layout different than any other backend screen?

File: src/pretix/base/models/auth.py [hunk 6]
```diff
@@ -593,6 +633,9 @@ def update_session_token(self):
         self.session_token = generate_session_token()
         self.save(update_fields=['session_token'])
 
+    def generate_email_verification_token(self):
```
Comment (raphaelm) [inline src/pretix/base/models/auth.py]: I think this is unused and broken?

Comment (raphaelm) [reply src/pretix/base/models/auth.py]: ah, it's used by pretixeu, but then its misisng an EmailVerificationTokenGenerator import

File: src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+            </button>
+            <a href="{{ cancel_url }}" class="btn btn-default">
+                {% trans "Cancel" %}
+            </a>
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html]: I think every other form has "Cancel" left and "Confirm" on the right, this one is reversed

File: src/pretix/base/models/auth.py (hunk 1 above)
Comment (luelista) [reply src/pretix/base/models/auth.py]: Done

File: src/pretix/control/views/user.py [hunk 8]
```diff
@@ -834,3 +820,130 @@ def get_object(self, queryset=None):
...
+            state=form.cleaned_data['new_email'],
+        )
+        return redirect(reverse('control:user.settings.email.confirm', kwargs={}))
+
+    def form_invalid(self, form):
+        messages.error(self.request, _('We could not save your changes. See below for details.'))
+        return super().form_invalid(form)
+
+
+class UserEmailConfirmView(FormView):
```
Comment (raphaelm) [inline src/pretix/control/views/user.py:918]: Why has this the style of our auth forms and not the new control dialog style? It's a little surprising in the user journey and I can't see that we use it in a logged-out state anywhere in this PR?

File: src/pretix/base/forms/auth.py [hunk 9]
```diff
@@ -307,3 +320,10 @@ def confirm_login_allowed(self, user: User):
                 self.error_messages['inactive'],
                 code='inactive',
             )
+
+
+class ConfirmationCodeForm(forms.Form):
+    code = forms.IntegerField(
+        label='',
```
Comment (raphaelm) [inline src/pretix/base/forms/auth.py]: Any reason not giving this a label? label='' is a little bit of a :middle_finger: to screen readers ;)

File: src/pretix/control/templates/pretixcontrol/user/settings.html [hunk 10]
```diff
@@ -3,6 +3,11 @@
 {% load bootstrap3 %}
 {% block title %}{% trans "Account settings" %}{% endblock %}
 {% block content %}
+{% if not user.is_verified %}
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/user/settings.html:6]: This will show up as an alert box for all existing users when we merge this, however, we have never sent them an email. That's not good?

I'm fine with having an alert for everyone, but I'd change the wording to a "your address is not confirmed, please confirm to secure your account" thing, and then add a button that sends an email with a token?

File: src/tests/control/test_user.py [hunk 11]
```diff
@@ -98,6 +98,20 @@ def test_change_email_no_duplicates(self):
         self.user = User.objects.get(pk=self.user.pk)
         assert self.user.email == 'dummy@dummy.dummy'
 
+
+class UserPasswordChangeTest(SoupTest):
```
Comment (raphaelm) [inline src/tests/control/test_user.py:151]: I think we need tests for the new email change form, its verification logic and edge cases, as well, as this is security-critical code.

File: src/pretix/control/views/user.py (hunk 8 above)
Comment (luelista) [reply src/pretix/control/views/user.py:918]: I think my idea was that this form would also be usable for a password-less login flow, or for email-based 2FA, if we would build something like that in the future.

But for this case it's probably better to use the new style.

File: src/pretix/base/forms/auth.py (hunk 9 above)
Comment (luelista) [reply src/pretix/base/forms/auth.py]: I think for the visual interface we don't need a label here, that's why I set it to empty string.  I'll add a label and mark it sr-only

Comment (luelista) [reply src/pretix/base/forms/auth.py]: Using the new dialog style, it looks okay with the label visible, so I just keep it visible instead

File: src/pretix/control/templates/pretixcontrol/user/settings.html (hunk 10 above)
Comment (luelista) [reply src/pretix/control/templates/pretixcontrol/user/settings.html:6]: Done

File: src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html [hunk 12]
```diff
@@ -0,0 +1,21 @@
+{% extends "pretixcontrol/auth/base.html" %}
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/auth/confirmation_code.html]: This template is unused and should be deleted

File: src/pretix/base/forms/auth.py [hunk 13]
```diff
@@ -307,3 +320,10 @@ def confirm_login_allowed(self, user: User):
                 self.error_messages['inactive'],
                 code='inactive',
             )
+
+
+class ConfirmationCodeForm(forms.Form):
+    code = forms.IntegerField(
+        label='Confirmation code',
```
Comment (raphaelm) [inline src/pretix/base/forms/auth.py]: Missing translation tags

File: src/pretix/base/forms/auth.py [hunk 14]
```diff
@@ -214,21 +214,38 @@ class PasswordRecoverForm(forms.Form):
...
     def __init__(self, user_id=None, *args, **kwargs):
-        self.user_id = user_id
-        super().__init__(*args, **kwargs)
+        initial = kwargs.pop('initial', {})
+        try:
+            self.user = User.objects.get(id=user_id)
+            initial['email'] = self.user.email
+        except User.DoesNotExist:
+            self.user = None
+        super().__init__(*args, initial=initial, **kwargs)
```
Comment (raphaelm) [inline src/pretix/base/forms/auth.py:248]: I know its not strictly new code but in what scenario makes it sense to render this form for a nonexistant user?

File: src/pretix/control/templates/pretixcontrol/user/settings.html [hunk 15]
```diff
@@ -3,8 +3,26 @@
...
+    <div class="alert alert-info">
+        <p>
+            {% blocktrans trimmed %}
+            Your email address is not confirmed yet. To secure your account, please confirm your email address by
+            clicking a confirmation link we send to your email address.
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/user/settings.html]: > To secure your account, please confirm your email address by clicking a confirmation link we send to your email address. 

but: 

> Please enter the confirmation code we sent to your email address foo@example.org.

:thinking: 

image

File: src/pretix/base/forms/auth.py (hunk 14 above)
Comment (luelista) [reply src/pretix/base/forms/auth.py:248]: Probably none, I did remove it, but then the test case that tests exactly that situation (tests.control.test_auth.PasswordRecoveryFormTest.test_recovery_unknown_user) failed, so I just put it back in. :D

File: src/pretix/control/templates/pretixcontrol/user/settings.html (hunk 15 above)
Comment (luelista) [reply src/pretix/control/templates/pretixcontrol/user/settings.html]: :D

File: src/pretix/base/forms/auth.py (hunk 14 above)
Comment (raphaelm) [reply src/pretix/base/forms/auth.py:248]: well… :shrug:

**Review by raphaelm — APPROVED** [review]
